src/tools/yarprobotstatepublisher/src/robotstatepublisher.cpp

```diff
@@ -252,16 +254,95 @@ void YARPRobotStatePublisherModule::onRead(yarp::rosmsg::sensor_msgs::JointState
         sizeOfTFFrames = model.getNrOfFrames();
     }
 
-    for (size_t frameIdx=0; frameIdx < sizeOfTFFrames; frameIdx++)
+    if (m_treeType == "SHALLOW")
     {
-        if(m_baseFrameIndex == frameIdx)    // skip self-tranform
-            continue;
-
-        iDynTree::Transform base_H_frame = m_kinDynComp.getRelativeTransform(m_baseFrameIndex, frameIdx);
-        iDynTree::toYarp(base_H_frame.asHomogeneousTransform(), m_buf4x4);
-        m_iframetrans->setTransform(m_tfPrefix + model.getFrameName(frameIdx),
-                                    m_tfPrefix + model.getFrameName(m_baseFrameIndex),
-                                    m_buf4x4);
+        // In shallow mode, we publish the position of each frame of the robot w.r.t. to the base frame of the robot
+        for (size_t frameIdx=0; frameIdx < sizeOfTFFrames; frameIdx++)
+        {
+            if(m_baseFrameIndex == frameIdx)    // skip self-tranform
+                continue;
+
+            iDynTree::Transform base_H_frame = m_kinDynComp.getRelativeTransform(m_baseFrameIndex, frameIdx);
+            iDynTree::toYarp(base_H_frame.asHomogeneousTransform(), m_buf4x4);
+            m_iframetrans->setTransform(m_tfPrefix + model.getFrameName(frameIdx),
+                                        m_tfPrefix + model.getFrameName(m_baseFrameIndex),
+                                        m_buf4x4);
+        }
+    }
+    else
+    {
+        // mode == DEEP
+        // In deep mode, we need to distinguish the following cases:
+        // For the frames that are frames of the link, we publish their location w.r.t. to their parent link
+        // For the additional frames, we publish their location w.r.t. to the frame of the link to which they are
+        // attached (note that this transform are actually constant)
+
+        // The traversal is the data structure that contains information on which link is parent of which other link,
+        // as in iDynTree the model is an undirected data structure
+        iDynTree::Traversal traversal;
+
+        // We generate a traversal using the base frame index
+        m_kinDynComp.model().computeFullTreeTraversal(traversal, m_baseFrameIndex);
+
+        bool setOk = false;
+
+        //Processing joints instead of links since it's easier this way to distinguish between static transform and non static ones
+        for (size_t jointIndex=0; jointIndex < model.getNrOfJoints(); jointIndex++)
+        {
+            auto currJoint = model.getJoint(jointIndex);
+            iDynTree::LinkIndex parentLinkIndex = currJoint->getFirstAttachedLink();
+            iDynTree::LinkIndex linkIndex = currJoint->getSecondAttachedLink();
+            iDynTree::Transform parentLink_H_link = m_kinDynComp.getRelativeTransform(parentLinkIndex, linkIndex);
```

Comment:
Notice that, depending on the choice of base, the parent link is not necessarily the first attached link (that's why the method is called ``getFirstAttachedLink`` and not ``getParentLink``). In order to get the parent and the child link of a given joint you should use [``getChildLinkIndexFromJointIndex``](https://robotology.github.io/idyntree/master/classiDynTree_1_1Traversal.html#af511b057b56553be5db56e34ca17c9a9) and [``getParentLinkIndexFromJointIndex``](https://robotology.github.io/idyntree/master/classiDynTree_1_1Traversal.html#a7c82fa00d69ac54fefdd89abcf1b2a5f) from ``Traversal``.

Reply:
Thanks a lot for the comment. I'm in vacation now, but I'll get back to it as soon as possible.